### PR TITLE
nsqd: add compression/decompression for messages

### LINKED ISSUE #1148: nsqd: enable compression/decompression pipeline from producer to consumer
NSQ currently supports snappy compression and decompression for TCP-protocol based producer and consumer. But it has some shortcomings.
* the memory and disk capacity nsqd used can not be decreased since all messages are stored just like what nsqd receives, i.e., the original messages.
* the in network bandwidth can not be decreased since the message are not compressed before it is send.
* they are just available for TCP protocol based producer and consumer.

I have made a proposal for this: https://docs.google.com/document/d/107yDH6pN8-b_i22AfvPkzdDU2yneOC-1lcUllpUmN4w/edit#heading=h.tla1oogr3wac

/cc @mreiferson @ploxiln

### PR DESCRIPTION
**Pull request by andyxning**
Fix #1148 

This PR adds compression/decompression pipeline for nsqd server.

Update:
* add proposal: https://docs.google.com/document/d/107yDH6pN8-b_i22AfvPkzdDU2yneOC-1lcUllpUmN4w/edit#heading=h.j5jbn91ci2bi

/cc @mreiferson @ploxiln

### CODE REVIEW

**Comment by andyxning** [comment]
Wait for @ploxiln 's re-design about nsqd storage backend.

File: nsqd/topic.go [hunk 1]
```diff
@@ -307,17 +307,12 @@ func (t *Topic) messagePump() {
...
-			// copy the message because each channel
-			// needs a unique instance but...
-			// fastpath to avoid copy if its the first channel
-			// (the topic already created the first copy)
-			if i > 0 {
-				chanMsg = NewMessage(msg.ID, msg.Body)
-				chanMsg.Timestamp = msg.Timestamp
-				chanMsg.deferred = msg.deferred
-			}
+		for _, channel := range chans {
```
Comment (andyxning) [inline nsqd/topic.go]: This is mainly used to avoid the race condition between topic's `PutMessage` method which is detected by the `go test -race`.

File: nsqd/protocol_v2.go [hunk 2]
```diff
@@ -126,7 +127,28 @@ func (p *protocolV2) SendMessage(client *clientV2, msg *Message) error {
...
+		adjustedBody, err = compress.DeflateDecompress(msg.Body)
+		if err != nil {
+			p.ctx.nsqd.logf(LOG_ERROR, "PROTOCOL(V2): decompress msg(%s) with deflate error", msg.ID)
+			return err
+		}
+		msg.deflateCompressed = false
+	}
+
+	msg.Body = adjustedBody
+	_, err = msg.WriteTo(buf)
```
Comment (mreiferson) [inline nsqd/protocol_v2.go]: I'm still confused why the changes here are necessary, the existing compression functionality already delivers compressed messages to consumers?

Comment (andyxning) [reply nsqd/protocol_v2.go]: Because the snappy or deflate compression is implemented by wraping the socket, i.e., nsqd will write original message over the wrapped socket and the compressed message will be actually be written to the wire. Reading is the same.

File: internal/compress/compress.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+		return nil, fmt.Errorf("error in decode snappy compressed message: %v", err)
+	}
+
+	return body, nil
+}
+
+func DeflateDecompress(compressedMsg []byte) ([]byte, error) {
+	fr := flate.NewReader(bytes.NewReader(compressedMsg))
+	defer fr.Close()
+	body, err := ioutil.ReadAll(fr)
```
Comment (andyxning) [inline internal/compress/compress.go]: mark

File: internal/compress/compress.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+	return body, nil
+}
+
+func DeflateDecompress(compressedMsg []byte) ([]byte, error) {
+	br := bytes.NewReader(compressedMsg)
+	fr := flate.NewReader(br)
+	defer fr.Close()
+	body, err := ioutil.ReadAll(fr)
+	if err != nil && !(err == io.ErrUnexpectedEOF && br.Len() == 0) {
```
Comment (andyxning) [inline internal/compress/compress.go]: new way

File: nsqd/protocol_v2.go (hunk 2 above)
Comment (mreiferson) [reply nsqd/protocol_v2.go]: Isn't that what we want to happen?

As we had discussed on https://github.com/nsqio/nsq/issues/1148#issuecomment-475809636, I feel like the way to accomplish what you want is for your producers and consumers to compress data.

It probably makes sense to get on the same page about the feature and whether or not it belongs in NSQ core before we discuss code and implementation.

Comment (andyxning) [reply nsqd/protocol_v2.go]: Yes. @mreiferson  You're right. I have also consider this pattern about compressing message in producer and decompressing it in consumer. However, On this way we need to do:
1、pass the compression algorithm information to the consumer for it to determine how to decompress it.
2、not all compression algorithm libs are supported by Golang/Python standard lib, for example snappy. In case to decompress the message correctly, consumer should know which compression algorithm the message use or the message can not be consumed correctly and will be requeued or even worse dropped directly once the `max_tries` is reached(I know that we can just print an error log and exit the consumer to indicate that something is wrong, but how about users set the `max_tries` to 1 and the message will be dropped next time it is consumed).

Comment (mreiferson) [reply nsqd/protocol_v2.go]: How you compress the messages is entirely up to you. In most cases you control both the producer and consumer — is there a need for more than one compression type to be used?

If multiple compression formats do need to be supported, you could also build some identifier into the message structure to indicate to consumers how to decompress.

Comment (ploxiln) [reply nsqd/protocol_v2.go]: If you have to transition from uncompressed messages to compressed messages, one way to do it is to start a new topic for the compressed messages ... so if you have "events" you might create another topic "events_gzip" and during the transition consume from both. Messages from "events_gzip" are gzipped, in this hypothetical case. But pick whatever is supported by both producer and consumer.

This scheme is nice because nsqd does not have to compress for storage and then decompress for delivery.

Comment (andyxning) [reply nsqd/protocol_v2.go]: > If multiple compression formats do need to be supported, you could also build some identifier into the message structure to indicate to consumers how to decompress.

I know this. But not every case we both control producer and consumer nor every producer need add compression byte info into a message. 

> If you have to transition from uncompressed messages to compressed messages, one way to do it is to start a new topic for the compressed messages ... so if you have "events" you might create another topic "events_gzip" and during the transition consume from both. Messages from "events_gzip" are gzipped, in this hypothetical case. But pick whatever is supported by both producer and consumer.

Yes. I agree with this. But what i want to express is to do this transparently to end users(producers and consumers).

File: internal/compress/compress.go (hunk 4 above)
Comment (waruqi) [reply internal/compress/compress.go]: Great! :+1:

**Comment by andyxning** [comment]
Let's wait for https://github.com/nsqio/nsq/pull/1170 to be merged, then it will be easy to add this feature. :)

**Comment by mreiferson** [comment]
@andyxning sorry I've been away for so long!

Context: the intended use case of the existing compression features of `nsqd` can best be described as "proxy compression", similar to using something like `nginx` and enabling gzip compression.

I noticed that you had a design doc linked in the original issue description with the following motivation:

> NSQ currently supports snappy compression and decompression for TCP-protocol based producer and consumer. But it has some shortcomings.
> 1. the memory and disk capacity nsqd used can not be decreased since all messages are stored just like what nsqd receives, i.e., the original messages.
> 2. the in-network bandwidth can not be decreased since the message are not compressed before it is send.
> 3. they are just available for TCP protocol based producer and consumer.

I think we should discuss these individually:

1. This seems like a reasonable request, at least for storage on disk, let's consider this for #1170 

2/3. When you publish messages via TCP protocol and enable compression on the connection, messages are compressed over the wire to `nsqd`. A reasonable addition could be to add Snappy and Deflate support _specifically to HTTP publish methods_. (There is no HTTP consumer, so it doesn't matter on the read side).

For this PR specifically, I would consider landing just the changes to HTTP publish methods to support `Content-Encoding`.

What do you think @andyxning and @ploxiln?

**Comment by jehiah** [comment]
👍 for adding `Content-Encoding` support on HTTP publish methods

**Comment by ploxiln** [comment]
👍 from me as well for HTTP Content-Encoding gzip/deflate (since those are already standard).

That probably won't satisfy @andyxning ... but I still think the way to go, when you really need good compression, is for the producer/consumer to do it. That way avoids nsqd doing extra compress and uncompress work for each message, and allows you to squeeze the most efficiency out of the system, if you need it.

**Comment by mreiferson** [comment]
Closing this, opened #1313 to track compression over HTTP pub.
